File: web/apps/app/lib/core/notes/__tests__/note-export.test.ts
import { describe, expect, it } from 'vitest';
import type { Note } from '@oriveo/shared';
import {
  buildNoteMarkdown,
  buildNoteMarkdownFilename,
  sanitizeNoteFilenamePart,
} from '../note-export';

function makeNote(overrides: Partial<Note>): Note {
  return {
    id: overrides.id ?? 'note-1',
    title: overrides.title ?? 'Answer title',
    titleSource: overrides.titleSource ?? 'manual',
    body: overrides.body ?? 'Body **markdown**',
    userNote: overrides.userNote,
    tags: overrides.tags ?? [],
    captureKind: overrides.captureKind ?? 'fullAnswer',
    sourceModelName: overrides.sourceModelName,
    sourceProviderName: overrides.sourceProviderName,
    sourcePrompt: overrides.sourcePrompt,
    createdAt: overrides.createdAt ?? '2026-06-10T08:30:00.000Z',
    updatedAt: overrides.updatedAt ?? '2026-06-10T08:30:00.000Z',
  };
}

describe('buildNoteMarkdown', () => {
  it('exports title, user note, body, tags and source footnote', () => {
    const markdown = buildNoteMarkdown(makeNote({
      userNote: 'Keep this for release notes',
      tags: ['release', 'web'],
      sourceModelName: 'GPT-5.4',
      sourceProviderName: 'OpenAI',
      sourcePrompt: 'Summarize the change',
    }), 'Untitled');

    expect(markdown).toContain('# Answer title');
    expect(markdown).toContain('> Keep this for release notes');
    expect(markdown).toContain('Body **markdown**');
    expect(markdown).toContain('Tags: release, web');
    expect(markdown).toContain('Source: GPT-5.4 - OpenAI - 2026-06-10');
    expect(markdown).toContain('Prompt: Summarize the change');
  });

  it('omits source for blank notes', () => {
    const markdown = buildNoteMarkdown(makeNote({ captureKind: 'blank', sourceModelName: undefined }), 'Untitled');

    expect(markdown).not.toContain('Source:');
  });
});

describe('note export filenames', () => {
  it('sanitizes illegal filename characters and truncates long text', () => {
    expect(sanitizeNoteFilenamePart('a/b\\c:d*e?f"g<h>i|j\nk')).toBe('a-b-c-d-e-f-g-h-i-j k');
    expect(sanitizeNoteFilenamePart('x'.repeat(120))).toHaveLength(80);
  });

  it('uses fallback title and note date', () => {
    expect(buildNoteMarkdownFilename(makeNote({ title: '' }), 'Untitled')).toBe('Untitled-2026-06-10.md');
  });
});

File: android/app/src/test/java/ai/oriveo/community/core/notes/NoteExportTest.kt
package ai.oriveo.community.core.notes

import ai.oriveo.community.core.model.Note
import ai.oriveo.community.core.model.NoteCaptureKind
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class NoteExportTest {

    private fun note(
        title: String = "My Title",
        body: String = "the body",
        userNote: String? = null,
        tags: List<String> = emptyList(),
        captureKind: NoteCaptureKind = NoteCaptureKind.FullAnswer,
        sourceModelName: String? = "GPT-5",
        sourceProviderName: String? = "OpenAI",
        sourcePrompt: String? = "the question",
        bodySnapshot: String? = "ORIGINAL SNAPSHOT",
    ) = Note(
        id = "11111111-1111-1111-8111-111111111111",
        title = title,
        body = body,
        userNote = userNote,
        tags = tags,
        captureKind = captureKind,
        sourceModelName = sourceModelName,
        sourceProviderName = sourceProviderName,
        sourcePrompt = sourcePrompt,
        bodySnapshot = bodySnapshot,
        createdAt = "2026-06-10T08:00:00Z",
        updatedAt = "2026-06-10T08:00:00Z",
    )

    @Test
    fun `markdown has expected field order and excludes snapshot`() {
        val md = NoteExport.buildNoteMarkdown(
            note(userNote = "why I saved it", tags = listOf("a", "b")),
            untitled = "Untitled",
        )
        assertTrue(md.startsWith("# My Title"))
        assertTrue(md.contains("> why I saved it"))
        assertTrue(md.contains("Tags: a, b"))
        assertTrue(md.contains("the body"))
        assertTrue(md.contains("Source: GPT-5 · OpenAI · 2026-06-10"))
        assertTrue(md.contains("Prompt: the question"))
        assertFalse(md.contains("ORIGINAL SNAPSHOT")) // bodySnapshot never exported
        assertTrue(md.endsWith("\n"))
    }

    @Test
    fun `blank capture omits source line`() {
        val md = NoteExport.buildNoteMarkdown(
            note(captureKind = NoteCaptureKind.Blank, sourcePrompt = null),
            untitled = "Untitled",
        )
        assertFalse(md.contains("Source:"))
    }

    @Test
    fun `empty title falls back to untitled`() {
        val md = NoteExport.buildNoteMarkdown(note(title = "  "), untitled = "Untitled note")
        assertTrue(md.startsWith("# Untitled note"))
    }

    @Test
    fun `filename sanitizes illegal chars and appends date`() {
        val name = NoteExport.buildNoteMarkdownFilename(note(title = "a/b:c*?\"<>|d"), untitled = "Untitled")
        assertFalse(name.contains("/"))
        assertFalse(name.contains(":"))
        assertTrue(name.endsWith("-2026-06-10.md"))
    }
}
